peps/pep-0667.rst

```diff
@@ -31,81 +30,166 @@ consistent regardless of threading or coroutines.
+Making the ``frame.f_locals`` attribute a write-through proxy
+-------------------------------------------------------------
+
+The Python 3.12 implementation of ``frame.f_locals`` returns a dictionary
+that is created on the fly from the array of local variables. The
+``PyFrame_LocalsToFast()`` C API is then called by debuggers and trace
+functions that want to write their changes back to the array (until
+Python 3.11, this API was called implicitly after every trace function
+invocation rather than being called explicitly by the trace functions).
+
 This can result in the array and dictionary getting out of sync with
-each other. Writes to the ``f_locals`` may not show up as
-modifications to local variables. Writes to local variables can get lost.
+each other. Writes to the ``f_locals`` frame attribute may not show up as
+modifications to local variables if ``PyFrame_LocalsToFast()`` is never
+called. Writes to local variables can get lost if a dictionary snapshot
+created before the variables were modified is written back to the frame
+(since *every* known variable stored in the snapshot is written back to
+the frame, even if the value stored on the frame had changed since the
+snapshot was taken).
 
 By making ``frame.f_locals`` return a view on the
 underlying frame, these problems go away. ``frame.f_locals`` is always in
 sync with the frame because it is a view of it, not a copy of it.
 
+Making the ``locals()`` builtin return independent snapshots
+------------------------------------------------------------
+
+:pep:`558` considered three potential options for standardising the behavior of the
+``locals()`` builtin in optimized scopes:
+
+* retain the historical behaviour of having each call to ``locals()`` on a given frame
+  update a single shared snapshot of the local variables
+* make ``locals()`` return write-through proxy instances (similar
+  to ``frame.f_locals``)
+* make ``locals()`` return genuinely independent snapshots so that
+  attempts to change the values of local variables via ``exec()``
+  would be *consistently* ignored rather than being accepted in some circumstances
+
+The last option was chosen as the one which could most easily be explained in the
+language reference, and memorised by users:
+
+* the ``locals()`` builtin gives an instantaneous snapshot of the local variables in
+  optimized scopes, and read/write access in other scopes; and
+* ``frame.f_locals`` gives read/write access to the local variables in all scopes,
+  including optimized scopes
+
+This approach allows the intent of a piece of code to be clearer than it would be if both
+APIs granted full read/write access in optimized scopes, even when write access wasn't
+needed or desired. For additional details on this design decision, refer to :pep:`558`,
+especially the :ref:`pep-558-motivation` section and :ref:`pep-558-exec-eval-impact`.
+
+This approach is not without its drawbacks, which are covered
+in the Backwards Compatibility section below.
```

Comment:
Missing pieces to get the PR merged:

- [ ] agreement from @markshannon that enough of the PR content has been moved over to PEP 558 for the rest to be acceptable (and what should be removed if the PR is still too big)
- [ ] confirmation from the SC that they're happy the missing impact assessment has now been added

Reply:
The SC is happy with the new text and has no suggestions or requests for any additional changes.

Reply:
Given the SC approval, I filed an issue for adding the hard deprecation to `PyEval_GetLocals()` for 3.14: https://github.com/python/cpython/issues/125170